controllers/druid/drain_historical.go

```diff
@@ -0,0 +1,191 @@
+		return err
+	} else {
+		msg := fmt.Sprintf("Successfully Patched replicas with %d", newReplica)
+		logger_drain_historical.Info(msg, "name", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+}
+
+func deployHistorical(sdk client.Client, m *v1alpha1.Druid, nodeSpec *v1alpha1.DruidNodeSpec, nodeSpecUniqueStr string, emitEvent EventEmitter, batchSize int32, baseUrl string) error {
+	// patch the updateStrategy with onDelete
+	err := patchUpdateStrategy(sdk, m, nodeSpec, onDelete, emitEvent)
+	if err != nil {
+		return err
+	}
+	// sts first scale up the pods by a batchsize and then drain the older pods
+	if m.Status.Historical.CurrentBatch == 0 {
+		m.Status.Historical = v1alpha1.HistoricalStatus{}
+		m.Status.Historical.Replica = nodeSpec.Replicas
+		m.Status.Historical.DecommissionedPods = []string{}
+
+		err := scaleStatefulSet(sdk, m, nodeSpec, onDelete, emitEvent, batchSize)
+		if err != nil {
+			return err
+		}
+
+		obj, err := readers.Get(context.TODO(), sdk, nodeSpecUniqueStr, m, func() object { return &appsv1.StatefulSet{} }, emitEvent)
+		if err != nil {
+			return err
+		}
+
+		if obj.(*appsv1.StatefulSet).Status.ReadyReplicas != obj.(*appsv1.StatefulSet).Status.CurrentReplicas {
+			msg := fmt.Sprintf("StatefulSet[%s] roll out is in progress CurrentReplicas[%d] != ReadyReplicas[%d]", nodeSpecUniqueStr, obj.(*appsv1.StatefulSet).Status.CurrentReplicas, obj.(*appsv1.StatefulSet).Status.ReadyReplicas)
+			logger_drain_historical.Info(msg, "name", m.Name, "namespace", m.Namespace)
+			return nil
+		}
+
+		m.Status.Historical.Replica = obj.(*appsv1.StatefulSet).Status.CurrentReplicas
+		m.Status.Historical.CurrentBatch = 0
+	}
+
+	originalReplicaCount := m.Status.Historical.Replica - batchSize
+
+	podList, err := readers.List(context.TODO(), sdk, m, makeLabelsForNodeSpec(nodeSpec, m, m.Name, nodeSpecUniqueStr), emitEvent, func() objectList { return &v1.PodList{} }, func(listObj runtime.Object) []object {
+		items := listObj.(*v1.PodList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	pvcLabels := map[string]string{
+		"druid_cr":          m.Name,
+		"nodeSpecUniqueStr": nodeSpecUniqueStr,
+		"component":         nodeSpec.NodeType,
+	}
+
+	pvcList, err := readers.List(context.TODO(), sdk, m, pvcLabels, emitEvent, func() objectList { return &v1.PersistentVolumeClaimList{} }, func(listObj runtime.Object) []object {
+		items := listObj.(*v1.PersistentVolumeClaimList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	podNames := getPodNames(podList)
+	//start draining the historical pod in batches and delete them
+	batchCount := originalReplicaCount/nodeSpec.DeploymentConfig.BatchSize + 1
+
+	//TODO add auth
+	d := InitClient(baseUrl, userName, os.Getenv("DRUID_PASSWORD"))
```

Comment:
You would have to add this env in the druid-operator deployment as well.

Reply:
Yes